python/azureml/score.py

```diff
@@ -0,0 +1,113 @@
+import os
+import logging
+import json
+import numpy as np
+import onnxruntime
+import transformers
+import torch
+
+# The pre process function take a question and a context, and generates the tensor inputs to the model:
+# - input_ids: the words in the question encoded as integers
+# - attention_mask: not used in this model
+# - token_type_ids: a list of 0s and 1s that distinguish between the words of the question and the words of the context
+# This function also returns the words contained in the question and the context, so that the answer can be decoded into a phrase. 
+def preprocess(question, context):
+    encoded_input = tokenizer(question, context)
+    tokens = tokenizer.convert_ids_to_tokens(encoded_input.input_ids)
+    return (encoded_input.input_ids, encoded_input.attention_mask, encoded_input.token_type_ids, tokens)
+
+# The post process function takes the list of tokens in the question and context, as well as the output of the 
+# model, the list of log probabilities for the choices of start and end of the answer, and maps it back to an
+# answer to the question that is asked of the context.
+def postprocess(tokens, start, end):
+    results = {}
+    answer_start = np.argmax(start)
+    answer_end = np.argmax(end)
+    if answer_end >= answer_start:
+        answer = tokens[answer_start]
+        for i in range(answer_start+1, answer_end+1):
+            if tokens[i][0:2] == "##":
+                answer += tokens[i][2:]
+            else:
+                answer += " " + tokens[i]
+        results['answer'] = answer.capitalize()
+    else:
+        results['error'] = "I am unable to find the answer to this question. Can you please ask another question?"
+    return results
+
+# Perform the one-off intialization for the prediction. The init code is run once when the endpoint is setup.
```

Comment:
Tiny nitpick: "initialization" is spelled incorrectly.

Reply:
Fixed